### DISCUSSION TITLE
Porting the AVR Transistortester to the LGT328P

### DISCUSSION THREAD

**Discussion by DurandA** [dbuezas/lgt8fx#112] [Show and tell]
I was thinking about porting the [AVR Transistortester](https://www.mikrocontroller.net/articles/AVR_Transistortester) to the LGT328P. In a nutshell, the AVR Transistortester is an open-source tool to identify components (resistor, capacitor, diodes...) and measure their characteristics. It is based on the ATmega processor. A [125 pages manual](https://github.com/svn2github/transistortester/blob/master/Doku/tags/english/ttester_eng112k.pdf) goes into the gory details to explain how it works. There is a gigantic [EEVblog topic](https://www.eevblog.com/forum/testgear/$20-lcr-esr-transistor-checker-project/) for in-depth discussion and [this link](https://github.com/blurpy/transistor-tester) nicely summarizes the mods that can be done to improve a popular Chinese version.

![9372908c-0b6f-4371-9132-139de73769b7](https://user-images.githubusercontent.com/3709037/109097936-9f49ba00-7720-11eb-96ef-b296b91129f8.jpg)

Using the LGT328P should give better accuracy thanks to the 12-bit ADC. Also the 32 Mhz clock will improve the reactivity on color displays.

I created [a repository](https://github.com/DurandA/transistor-tester-lgt328p) with the latest m-firmware. The [official source](https://www.mikrocontroller.net/svnbrowser/transistortester/Software/Markus/) is a SVN repo with the firmware distributed as tarball inside the repo which is cumbersome. I have no experience with the LGT328P so I would be interested in your experience about the compatibility with the ATmega. Do you think the port will just be a few adjustments to the peripherals (e.g. ADC registers) or it could be more complex?

**Comment by DurandA** [comment]
Early support to the LGT328P infrastructure is ongoing.

I added 24 and 32 Mhz wait functions but need to confirm that [these figures](https://github.com/DurandA/transistor-tester-lgt328p/blob/ef3754936aeb8f7d5ace202dffcdbb76e571e075/wait.S#L198-L199) are still valid (I asked @LaZsolt [here](https://github.com/dbuezas/lgt8fx/issues/18)):
 * rcall needs 3 clock cycles
 * ret needs 4 clock cycles

If true, then the waiting functions can be kept intact since these are based on nops.

Next step is to update the [ADC](https://github.com/DurandA/transistor-tester-lgt328p/blob/lgt328p/ADC.c). Is the ADC supposed to work when using the ATmega 328P registers?

**Reply by LaZsolt** [reply]
In LGT8F328x Some instructions are faster than in ATmega.
**rcall** needs only 1 clock cycle, so need to insert 2 more nops
**ret** needs 2 clock cycles, so need to insert 2 more nops too
for the same delaying than ATmega.

You need to recalculate the number of nops on every different frequencies.
If you insert 4 nops into the delay core you will get the same delay timings.

Here is a small trinck to make your code a bit shorter. Every groups of 3 nops could be replaced this:
```
brcc .+0
brcs .+0
```
This trick works when using LGT8Fx MCU

Edit: instructions brcc + brcs

**Reply by dbuezas** [reply]
This table may be useful https://docs.google.com/spreadsheets/d/1EzwMkWOIMNDqnjpbzuchsLx5Zq_j927tvAPgvmSuP6M/edit?usp=sharing

**Reply by DurandA** [reply]
This is really appreciated, thanks!

**Comment by LaZsolt** [comment]
@DurandA:
> Is the ADC supposed to work when using the ATmega 328P registers?

No. The ADC of LGT8F328P differs from ATmega328P ADC. Even not the same in LGT8F328**D**
- In LGT8F328P you may select 3 different reference voltages, while in LGT8F328**D** has 2 different reference voltages. But LGT8F328**D** reference voltage precisity much lower than P version.
- The resolution of LGT8f328x's ADC are 12 bit.
- Has dynamic calibration, but you must use it for better accuracy.
and more...

**Reply by DurandA** [reply]
> No. The ADC of LGT8F328P differs from ATmega328P ADC. Even not the same in LGT8F328**D**

I asked because `ADCSRA`/`ADCSRB` registers addresses are the same so I assumed it would work without the extra features.

I had a closer look at the ADC prescaler on the datasheet:
* "The highest resolution at sample rates up 500KSPS"
* "By default, the successive approximation circuitry requires an 300KHz To 3MHz The input clock to get maximum resolution."

If I understand this correctly, to get the best resolution the prescaler should always result in a 300kHz to 3MHz ADC clock (compared to the ATmega 328P that needs an ADC clock of 50kHz to 200kHz). The 500KSPS figure looks like a main 32Mhz clock and a prescaler of /64.  I'm not sure about the upper 3MHz ADC clock "to get maximum resolution" since with a prescaler of /16 the ADC clock would run at 2Mhz and the sample rate would be 2MSPS. Maybe it is a translation issue.

Should I chose the lowest prescaler that results in an ADC clock >= 300kHz for best resolution?

Thank you for your time.

**Reply by dbuezas** [reply]
> with a prescaler of /16 the ADC clock would run at 2Mhz and the sample rate would be 2MSPS

The ADC takes 22 cycles to compleate a reading, so 32Mhz/16/22 = 90kS/s
It is supposed to take 15.5 cycles (datasheet), but experimentally me and others found it needs 22.


Relevant: https://github.com/dbuezas/arduino-web-oscilloscope/blob/master/arduino/src/adc.h#L57

Using it at 32Mhz cpu clock and in 8 bit precision mode, we've found that:
* Prescaler=2 : cannot read more than 3v
* prescaler=4 (363kS/s) : works perfectly with 8 bits readings.
* rest: same as 4.

For 12 bits, I would assume that a prescaler of 8 should be fine, but 4 may also work.

**Reply by DurandA** [reply]
> The ADC takes 22 cycles to compleate a reading, so 32Mhz/16/22 = 90kS/s

Thanks for the clarification.

> Using it at 32Mhz cpu clock and in 8 bit precision mode, we've found that:
> 
>     * Prescaler=2 : cannot read more than 3v
> 
>     * prescaler=4 (363kS/s) : works perfectly with 8 bits readings.

It makes sense because that the readings are not correct with a prescaler of 2 as the ADC clock is way over the 300KHz to 3Mhz range (=16Mhz).

**Reply by DurandA** [reply]
I played a bit with the LGT328P ADC and it is quite compatible with the ATmega when using default register values. For example, assuming the prescaler was adjusted, the following will work on both the LGT and the ATmega:
```
  ADCSRA = (1 << ADEN) | ADC_CLOCK_DIV;
  ADMUX  = 0b01000000;

  ADCSRA |= (1 << ADSC);
  while (ADCSRA & (1 << ADSC)) {;;}
  int adc = ADC;
```

The icing on the cake is that the adc value on the LGT will be 12-bit without adjustments.

**Comment by DurandA** [comment]
I am really pleasantly surprised by this chip, the compatibility the the ATmega328P is much better than I anticipated:
* Digital GPIO works fine without modification. That means that bit-banged protocols are compatible as long as the delay are compensated. I was able to use the bit-banged UART protocol from the m-firmware.
* ADC works fine with default settings with adjusted prescaler. The m-firmware seems to use only AVCC and the ATmega 1.1V reference. Some small tuning is required as the LGT has a 1.024 reference.
* I like the fact that there is no fuses and that the clock source can be changed at runtime. Using an external crystal works just fine.
* I'm sure the extra voltage references, the DAC and the differential analog compactor could enable new kinds of measurement. This is however not something I will try as it requires a lot of analog/EE knowledge.

Do you know how much compatible are the hardware implementation of TWI/I2C, UART and SPI compared to the ATmega? I did not have the time to compare the datasheets yet. They are not strictly required by the m-firmware since there are bit-banged version of each of them. For most existing boards, they are unusable anyway as the required pins are already used for other purposes.

**Reply by LaZsolt** [reply]
To the best of my knowledge to controlling TWI and UART has no difference with ATmega328.
But SPI is totally different than ATmega. It has 4 bytes buffer which cause difficulties. But if you could use the SPI library, it will work without problems.

**Reply by DurandA** [reply]
Thanks for the precious information.

There will be no updates in the coming days as I am working on a modified PCB for this project. I will need to wait a few weeks to get every components and assemble a board before I can debug the firmware.

**Reply by DurandA** [reply]
Quick update: I forked an [EasyEDA version](https://oshwlab.com/wegi1/m328-transistor-tester) of the M328 transistor tester:

![transistor-tester](https://user-images.githubusercontent.com/3709037/109863324-cc86f280-7c61-11eb-9449-79e8a36f813f.png)

This should be compatible with both the LGT328P and ATmega329P(B) (minus the ISP header) for direct comparison. I am now waiting for the PCB and components which could take a while.

**Reply by youxiaojie** [reply]
how about the new version of tester?

**Reply by DurandA** [reply]
I received the PCB and have all the spare parts except the LGT328P chips. I don't want to transplant the MCU from my only board so I will wait a little more.

**Reply by youxiaojie** [reply]
good site! open source hard ware.
I also want to make arduino by 328d/32pins and 328p/48pings to get more gpios
the chip is very chip, 2RMB for 328d to 4RMB 328p/48pins

**Comment by DurandA** [comment]
I received the components and started the assembly. Unfortunately, 0.1% resistors in thru-hole are not stocked.

![transistor-tester](https://user-images.githubusercontent.com/3709037/115974797-ac87f700-a55f-11eb-80da-73571d9b8321.jpg)

I will publish more details as I test the firmware.

**Reply by dbuezas** [reply]
Did you succeed? I'm looking forward to your results! :)

**Reply by DurandA** [reply]
I had to put on hold all my personal projects in order to write a university thesis. I am looking forward to continue and finish this project. :)

**Reply by dbuezas** [reply]
I assume you must have finished your thesis by now, so:
Congratulations! :)
